### ISSUE TITLE
Query Cancellation 2.0

### ISSUE THREAD

**Issue by camsaul** [metabase/metabase#12116]
Here's my quick proposal for query cancellation 2.0 for whoever ends up implementing it (not me since I will be OOO):

*  We can have the FE client generate UUIDs and send them along with the queries as @paulrosenzweig suggested. How this happens is up for debate. Paul suggested passing it in as an HTTP header which sounds reasonable.
*  We should add a new endpoint that the FE client can use to cancel request(s) with the given UUID(s)
*  Since we can't be sure the server that is running the query will be the same one getting the cancel request we'll need to make sure cancellation can work across multiple instances. So something like a new `query_cancellation` table indexed on the UUID might be the best solution. 
      *  The API endpoint can add an entry to the `query_cancelation` table to signify that the query with that UUID should be canceled
      *  To keep this table from getting out of control huge we should clear out old entries, I would just go ahead and do it at the same time we add new ones. 
            *  Perhaps anything older than 24 hours? 
            *  To do this efficiently we will want an index on the `created_at` column as well
*  The streaming response code in `metabase.async.streaming-response` needs to be able to figure out what UUID is associated with current request. 
    *  Assuming the cancellation UUID is passed in as an HTTP header, `metabase.async.streaming-response/respond` will be able to pull this value out of the Ring `request-map` which it gets as a parameter.
*  The streaming response code should periodically check whether a cancellation entry is present in the `query_cancellation` table for its requests' UUID. You could either do this with one loop per-request or a shared loop:
      *  the one-loop-per-request pattern is simpler
            *  To avoid too much load on the application DB I think having this loop check every 30 seconds or so should be sufficient, but we'll have to think about a good trade-off between too many pointless application DB queries and not canceling slow queries as fast as desired
            *  Needless to say the check-for-cancellation loop should stop running once the query completes. Once `finished-chan` gets a value the request is complete, so if you do a `core.async` loop you can use `alts!` to check that. `finished-chan` is a promise channel that guaranteed to get a value when the query completes, is canceled, or throws an Exception.
      *  One global loop is more efficient but a little trickier
            *  One way to do it would be to store a map of `UUID->future` in memory, and have the single loop fetch all recent `query_cancellation` entries, then `future-cancel` all relevant tasks.
                  *  Should remove `UUID->future` entries when tasks complete or are canceled so we don't hold them in memory forever
      *  A hybrid approach that uses one loop-per-request but a shared (TTL-cached?) function to fetch recent cancellation entries from the DB could be a best-of-both-worlds approach 🤷‍♂ 
* To actually perform query cancellation, all you need to do is send a message to `canceled-chan` and `future-cancel` the task running the query in `metabase.async.streaming-response` e.g.
 
###### Current Code

 ```clj
(defn- do-f-async [f ^OutputStream os finished-chan]
  (let [canceled-chan (a/promise-chan)
        task          (bound-fn []
                        ...)]
    (.submit (thread-pool/thread-pool) ^Runnable task)
    nil))
```

###### Example with query cancelation:

```clj
(defn- do-f-async [f ^OutputStream os finished-chan cancellation-uuid]
  (let [canceled-chan (a/promise-chan)
        task          (bound-fn []
                        ...)
        futur         (.submit (thread-pool/thread-pool) ^Runnable task)]
    (a/go-loop []
      (let [check-cancellation-timeout-chan (a/timeout 30000)
            [_ port]                        (a/alts! [check-cancellation-timeout-chan finished-chan])]
        (when (= port check-cancellation-timeout-chan)
          (if (cancel-request? cancellation-uuid)
            (do
              (a/>! canceled-chan ::cancel)
              (future-cancel futur))
            (recur)))))
    nil))
```

Exact implementation details are left as an exercise to the reader. This is just one idea for going about implementation.

**Comment by camsaul** [comment]
the backend implementation should be pretty straightforward you should be able to do almost everything in the metabase.async.streaming-response namespace (and only 2 or 3 functions would need to be tweaked). The only other thing you'd need to add are the new API endpoint and the new app DB table.

**Comment by sbelak** [comment]
I think the localised (one loop per request) approach is better as it keeps the code together & has less coupling (and with memoization can be just as efficient and can further decouple the caching/fetching logic). 

Is information which queries got canceled (the most) something that would be useful for admins in audit view?

**Comment by flyingmachine** [comment]
## Additional Context?

Could I get a little more context for this feature?

* What are the users' needs? Is this about managing performance?
* We want users to be able to cancel requests - what does that look like? Clicking a button?
* Do we want to always cancel requests when a user navigates away from a view that initiated queries?
* What is the relationship between an in-progress query and an open HTTP connection?
* Is it possible for User A to initiate a query and for User B to navigate to a view of that in-progress query?

## What's the model?

From the description it sounds like we're talking about a lightweight job management system. Jobs are distributed across MB servers and tracked in a table. Job management commands are conveyed via the table. Is that a useful way of thinking about this, and if it is does it make sense to bring in other tools from job management systems, like a job management UI? I realize that would increase the scope and is probably undesirable; the question is meant to help me understand the nature of the problem.

Would it be possible to use Postgres's [async notifications](https://www.postgresql.org/docs/9.3/libpq-notify.html) instead of polling a table?

## Failure handling

* What are the failure cases?
* Is it possible for a future to get cancelled but for the DB to continue executing the query?
* More generally, can jobs get orphaned and is it possible to know about that or handle it?
* How will we know something's gone wrong?
* How will we repro issues?
* How will users send us troubleshooting info?

**Comment by paulrosenzweig** [comment]
Some feature context:

Users will often realize an error while a slow query is running, edit the query, and start it again. If the original query had a bug that made it really slow, that query can keep running for hours. Do that a few times as you iterate on the query and you can hang the db.

Users can cancel queries by clicking a button or a few implicit ways. Previously we tried handle that by having the client cancel HTTP requests. The new way shouldn't rely on request cancellation to make sure it reliably works for all setups.

A useful trick for testing cancelation in postgres: `select pg_sleep(10);` takes 10 seconds to return. By tailing the postgres logs you can see when the query was successfully canceled. 

---

> What is the relationship between an in-progress query and an open HTTP connection?

An in-progress query should always have an pending HTTP request. I'm guessing that's why we originally wanted request cancelation to trigger query cancelation.

> Is it possible for User A to initiate a query and for User B to navigate to a view of that in-progress query?

Not currently

**Comment by sbelak** [comment]
My understanding (or view) on resilience/failure cases is that there's a hierarchy: manual cancelation > implicit cancelations and that it's not the end of the world if a cancel request sometimes even slips through the cracks as long as pressing "cancel" again does the job. 

I remember from back when we first attempted cancelation that how/when the DB actually cancels the query somewhat differs from DB to DB. One thing we should definitely test/check is what happens on the DB side if we think the request has been canceled, but it's in fact not (can this lead to some sort of exhaustion).

**Comment by mazameli** [comment]
> What are the users' needs? Is this about managing performance?

I think it's about keeping the DW from keeling over. In the UI, it's about having a button to make the loading spinner go away so I can edit the query.

> We want users to be able to cancel requests - what does that look like? Clicking a button?

Yes, there's a cancel button in the UI when a query is running in the GUI or SQL modes, but not when viewing dashboards.

> Do we want to always cancel requests when a user navigates away from a view that initiated queries?

That seems to make sense to me, unless/until we someday add UX for initiating a query and then being notified when it's finished.

**Comment by tlrobinson** [comment]
> Previously we tried handle that by having the client cancel HTTP requests. The new way shouldn't rely on request cancellation to make sure it reliably works for all setups.

Is there any context on why this sometimes didn't work?

Cancelling when the HTTP connection is closed seems much simpler than having a bunch of state that everyone has to keep track of. It also handles implicit cancellation automatically, i.e. when the browser window is closed.

The one thing this solution could get us is deduplicating identical concurrent requests, i.e. if multiple users load the same dashboard at the same time. I think if we implemented this design it might also make sense to also implement that (though then we might need a reference count or something, so one user can't cancel everyone else's request too)

**Comment by flyingmachine** [comment]
What @camsaul shared in slack:

> Our old approach for query cancelation was to hold the OutputStream open and write new lines to it every second and if it threw a connection closed exception then cancel. That approach doesn't work because 1) writing new lines doesn't work for CSV/XLSX downloads and 2) holding open the OutputStream consumes a limited Jetty resource handle somewhere so you can exhaust all Jetty handles and get the server in an unresponsive state if you have a lot of pending queries. 
> If we are writing the response and hit a query canceled EofException then we still do cancel it but this doesn't happen until there is a response to write and we would like to cancel long-running requests right away rather than waiting until they finish.

**Comment by tlrobinson** [comment]
FWIW it appears Node.js (and others, I'm sure) is able to detect closed connections without sending bytes (I tried it). It might be worth evaluating other Ring servers if Jetty is also wasting a lot of memory on open connections, e.x. `http-kit` or maybe `ring-netty-adapter` https://github.com/ring-clojure/ring/wiki/Third-Party-Libraries. The so-called ["c10k problem"](https://en.wikipedia.org/wiki/C10k_problem) was solved like 10 years ago now...

If there's really no other workaround then this is fine.

It would be nice if we while we're putting in the effort we implemented some other features it would unlock:

1. deduplicating identical concurrent requests, as mentioned in my comment above (which could drastically reduce load on highly trafficked dashboards). caching only applies if the query has completed before the next one is made, I think?
2. a panel in the admin troubleshooting section showing pending long running queries, and allowing the admin to cancel them
